### PR TITLE
chore: add `config.nims` with nim defaults & chore tasks

### PR DESCRIPTION
**Pull request by ZoomRmc**
This PR adds a `config.nims` that provides:
- A few compiler defaults (some of them based on PR #20) .
- A Nim task `test` to run all the units in the repo as executables, with an assumption it runs tests. Task runs with `nim test`.
  This task is a proposed replacement for the shell scripts from PR #20.
  Main benefits over Bash: cross-platform, no additional dependencies.
- A Nim task `prettyfy` ro run `nimpretty` on all *.nim files in the repository. Task runs with `nim prettyfy`
  This task is a proposed replacement for the shell script from PR #19.

List of compiler defaults:
```nim
--gc:arc
--checks:on
--assertions:on
--spellSuggest
--styleCheck:error
```

### CODE REVIEW

File: config.nims [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      yield path
+
+############ Tasks
+task test, "Test everything":
+  --hints: off
+  var failedUnits: seq[string]
+
+  for path in modules():
+    if not path.isIgnored() and path.endsWith(".nim"):
+      echo &"Testing {path}:"
```
Comment (dlesnoff) [inline config.nims:37]: This is fine for now, but I will probably want to reduce the verbosity as soon as we have more than a few dozen of scripts.

**Review by dlesnoff — CHANGES_REQUESTED** [review]
Apart from being cross-platform, I prefer Nimscript configuration file that provides an example for people wanting to set up a Nim project. I do not want to depend on Nimble, nor on Atlas.

File: config.nims [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# Nim compilation defaults for the whole reposiotry
+--gc: arc
```
[the remark this answers is not in the document]
Comment (ZoomRmc) [reply config.nims:2]: I was looking at it from the perspective that compilation happens only for testing. Accounting for the expected scope of individual units, possible leaking some memory seems like not such a big deal. Orc, on the other hand, is a bit less stable and usually sees more compiler bugs. Other mm modes are legacy (or soon-to-be).

I don't mind changing defaults to orc.

File: config.nims [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    if not path.isIgnored() and path.endsWith(".nim"):
+      yield path
+
+############ Tasks
+task test, "Test everything":
+  --hints: off
+  var failedUnits: seq[string]
+
+  for path in modules():
+    if not path.isIgnored() and path.endsWith(".nim"):
```
[the remark this answers is not in the document]
Comment (ZoomRmc) [reply config.nims:36]: Oops, thanks!

File: config.nims (hunk 1 above)
Comment (ZoomRmc) [reply config.nims:36]: This line just delineates the `unittest` output. With current config when everything passes smoothly outputs for different modules don't have any other separation.

File: config.nims (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZoomRmc) [reply config.nims:2]: Regarding documenting it, let's see if anyone ever has any problem with arc here first :)
`config.nims` is pretty self-explanatory anyway.

**Review by dlesnoff — COMMENTED** [review]
I think that this is fine, but I wonder how I will merge it with Github actions from previous PRs. I need to ponder about it a bit.

**Comment by vil02** [comment]
@ZoomRmc The benefits are clear - I fully support your idea.

Regarding `nim prettyfy` - it does not reformat the files at my end.

**Comment by ZoomRmc** [comment]
> @ZoomRmc The benefits are clear - I fully support your idea.

Thanks. 

> Regarding `nim prettyfy` - it does not reformat the files at my end.

Works on my end. It's rather silent, though. Did you reset your working tree to main? It should make the changes in the dynamic_programming dir.

**Comment by vil02** [comment]
> > Regarding `nim prettyfy` - it does not reformat the files at my end.
> 
> Works on my end. It's rather silent, though. Did you reset your working tree to main? It should make the changes in the dynamic_programming dir.

I have introduced some incorrect formatting and it was not fixed. I have nimpretty 0.2 and  Nim Compiler Version 1.6.12 [Linux: amd64].

**Comment by ZoomRmc** [comment]
> I have introduced some incorrect formatting and it was not fixed. I have nimpretty 0.2 and Nim Compiler Version 1.6.12 [Linux: amd64].

Well did the nimpretty run or not with `nim prettyfy`? This task does little besides that. Nimpretty is not super-reliable and definitely needs a bit more love.

**Review by dlesnoff — APPROVED** [review]

**Comment by vil02** [comment]
> > I have introduced some incorrect formatting and it was not fixed. I have nimpretty 0.2 and Nim Compiler Version 1.6.12 [Linux: amd64].
> 
> Well did the nimpretty run or not with `nim prettyfy`? This task does little besides that. Nimpretty is not super-reliable and definitely needs a bit more love.

Now I could check it again: it turned out that introducing some big number of empty lines is not considered as incorrect format - sorry for the confusion, it works well.

**Comment by vil02** [comment]
@ZoomRmc I am having another issue. When running the command `nim test`, I get the error (both locally and [here](https://github.com/vil02/Nim/actions/runs/5159968029/jobs/9295525103?pr=3)):
`/home/runner/.choosenim/toolchains/nim-1.6.12/lib/pure/parseutils.nim(426, 21) Error: system module needs: nimDestroyAndDispose`

When I do not use `--gc: arc`, everything works well.

**Comment by ZoomRmc** [comment]
@vil02 good catch, thanks.

That's a compiler error, it's fixed on devel (not yet sure at which point though). For mitigation we can move the `gc` switch from `config.nims` to `nim.cfg`.

**Comment by vil02** [comment]
@ZoomRmc Thanks, cf. 9c9c3831b1eb85272fb1a19fd45b9b2eca45408a in #20.

**Pull request merged**